[AMDGPU] Add target intrinsic for s_prefetch_data

rampitec opened the pull request.

rampitec committed:
[AMDGPU] Add target intrinsic for s_prefetch_data

shiltian commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -19489,6 +19489,12 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
             EmitScalarExpr(E->getArg(2)), EmitScalarExpr(E->getArg(3))});
   }
+  case AMDGPU::BI__builtin_amdgcn_s_prefetch_data: {
+    llvm::Value *Ptr = EmitScalarExpr(E->getArg(0));
```
can use `emitBuiltinWithOneOverloadedType<1>(...)`

rampitec replied:
I does not have a return value, so cannot.

shiltian replied:
`emitBuiltinWithOneOverloadedType` doesn't have to have a return value

rampitec replied:
Hm, ok, The description of the function is misleading. Done.

rampitec committed:
Fix legalization of S_PREFETCH

$soffset can be selected to SGPR_NULL and MRI.getRegClass()
expects a virtual register.

rampitec committed:
Use emitBuiltinWithOneOverloadedType()

shiltian commented on llvm/test/CodeGen/AMDGPU/llvm.amdgcn.s.prefetch.data.ll:
```diff
@@ -0,0 +1,136 @@
+define amdgpu_ps void @prefetch_data_sgpr_base_imm_len_global(ptr addrspace(1) inreg %ptr) {
+; GCN-LABEL: prefetch_data_sgpr_base_imm_len_global:
+; GCN:       ; %bb.0: ; %entry
+; GCN-NEXT:    s_prefetch_data s[0:1], 0x0, null, 31
```
Why does this not have `v_readfirstlane_b32`? `isFlatGlobalAddrSpace` returns true for AS1 as well?

shiltian replied:
That makes sense. Thanks for explanation.

arsenm commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -2689,6 +2689,12 @@ def int_amdgcn_global_load_tr_b128 : AMDGPULoadIntrinsic<global_ptr_ty>;
   DefaultAttrsIntrinsic<[llvm_i32_ty], [], [NoUndef<RetIndex>, IntrNoMem, IntrSpeculatable]>;
 
+def int_amdgcn_s_prefetch_data :
+  Intrinsic<[], [llvm_anyptr_ty, llvm_i32_ty],
```
Comment the operands? How is this different from the regular llvm.prefetch?

rampitec replied:
Added comments. I will have to guard the regular llvm.prefetch, the instruction turns to be unsafe with invalid address. Besides target intrinsic can take length operand.

rampitec committed:
Commented intrinsic operands
